### PR TITLE
Add generate_series function

### PR DESCRIPTION
**Pull request by divanik**
### Changelog category:
New Feature

### Changelog entry:
Add generate_series as a table function. This function generates table with an arithmetic progression with natural numbers.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit f91feb0dcb405df80f317f456372c7374f2c75ee with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59390/f91feb0dcb405df80f317f456372c7374f2c75ee/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by qoega** [comment]
- [ ] tests
- [ ] documentation
- [ ] cla

**Comment by alexey-milovidov** [comment]
@divanik, it slowed down, according to the performance tests.

File: docs/en/sql-reference/table-functions/generate_series.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+---
+slug: /en/sql-reference/table-functions/generate_series
+sidebar_position: 146
+sidebar_label: generate_series
+---
+
+# generate_series
```
Comment (rschu1ze) [inline docs/en/sql-reference/table-functions/generate_series.md:7]: For consistency with `generateRandom`, this should really be called `generateSeries`.

File: docs/en/sql-reference/table-functions/generate_series.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+`generate_series(START, STOP, STEP)` - Returns a table with the single ‘generate_series’ column (UInt64) that contains integers from start to stop inclusively with spacing between values given by STEP. 
+
+The following queries return tables with the same content but different column names:
+
+``` sql
+SELECT * FROM numbers(10, 5);
+SELECT * FROM generate_series(10, 14);
+```
+
+And the following queries return tables with the same content but different column names (but the second option is more efficient):
```
Comment (rschu1ze) [inline docs/en/sql-reference/table-functions/generate_series.md:20]: Existing table function `numbers` has overloads
- `numbers(N)` - "generate numbers 0 to N-1"
- `numbers(N, M)` - "generate numbers N to N + M - 1".

Why don't we add a third overload
- `numbers(N, M, S) - "generate numbers N to N + M - 1 with step size S"

? That would imho be more user-friendly than adding a new and unrelated table function.

File: docs/en/sql-reference/table-functions/generate_series.md (hunk 1 above)
Comment (alesapin) [reply docs/en/sql-reference/table-functions/generate_series.md:7]: I think we shoud add `alias` for both versions.

File: CMakeLists.txt
Comment (divanik) [inline CMakeLists.txt:1]: Must be reverted after fix of the build with a sanitizer

File: src/Common/iota.h [hunk 3]
```diff
@@ -31,4 +31,13 @@ extern template void iota(UInt64 * begin, size_t count, UInt64 first_value);
 #if defined(OS_DARWIN)
 extern template void iota(size_t * begin, size_t count, size_t first_value);
 #endif
+
+template <iota_supported_types T> void iota_with_step(T * begin, size_t count, T first_value, T step);
```
Comment (alesapin) [inline src/Common/iota.h:35]: why snake_case since it's our function? `iotaWithStep`?

File: src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp [hunk 4]
```diff
@@ -23,19 +29,34 @@ extern const int TOO_MANY_ROWS;
 namespace
 {
 
+template <iota_supported_types T>
+inline void iota_with_step_optimized(T * begin, size_t count, T first_value, T step)
```
Comment (alesapin) [inline src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp:33]: Also should be `iotaWithStepOptimized`

File: src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp [hunk 5]
```diff
@@ -65,43 +88,57 @@ class NumbersSource : public ISource
...
-    if (r.right.isPositiveInfinity())
-        return static_cast<UInt128>(std::numeric_limits<uint64_t>::max()) - r.left.get<UInt64>() + r.left_included;
+};
 
-    size = static_cast<UInt128>(r.right.get<UInt64>()) - r.left.get<UInt64>() + 1;
+using RangesWithStep = std::vector<RangeWithStep>;
 
-    if (!r.left_included)
-        size--;
+std::optional<RangeWithStep> stepped_range_from_range(const Range & r, UInt64 step, UInt64 remainder)
```
Comment (alesapin) [inline src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp:105]:
```suggestion
std::optional<RangeWithStep> steppedRangeFromRange(const Range & r, UInt64 step, UInt64 remainder)
```

File: src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp [hunk 6]
```diff
@@ -65,43 +88,57 @@ class NumbersSource : public ISource
...
+        total_size += r.size;
     }
     return total_size;
 };
 
 /// Generate numbers according to ranges.
 /// Numbers generated is ordered in one stream.
 /// Notice that we will not generate additional numbers out of ranges.
-class NumbersRangedSource : public ISource
+class [[maybe_unused]] NumbersRangedSource : public ISource
```
Comment (alesapin) [inline src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp:141]: What is the motivation for this change?

File: src/Storages/System/StorageSystemNumbers.cpp [hunk 7]
```diff
@@ -11,17 +11,22 @@
 #include <Processors/Sources/NullSource.h>
 #include <QueryPipeline/Pipe.h>
 #include <Storages/SelectQueryInfo.h>
+#include <Common/logger_useful.h>
```
Comment (alesapin) [inline src/Storages/System/StorageSystemNumbers.cpp:14]: redundant header

File: src/Storages/System/StorageSystemNumbers.cpp [hunk 8]
```diff
@@ -11,17 +11,22 @@
...
+    std::optional<UInt64> limit_,
+    UInt64 offset_,
+    UInt64 step_)
+    : IStorage(table_id), multithreaded(multithreaded_), limit(limit_), offset(offset_), column_name(column_name_), step(step_)
 {
     StorageInMemoryMetadata storage_metadata;
-    /// This column doesn't have a comment, because otherwise it will be added to all the tables which were created via
-    /// CREATE TABLE test as numbers(5)
-    storage_metadata.setColumns(ColumnsDescription({{"number", std::make_shared<DataTypeUInt64>()}}));
+    storage_metadata.setColumns(ColumnsDescription({{column_name_, std::make_shared<DataTypeUInt64>()}}));
```
Comment (alesapin) [inline src/Storages/System/StorageSystemNumbers.cpp:29]: return comment back?

File: src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp [hunk 9]
```diff
@@ -419,80 +474,100 @@ Pipe ReadFromSystemNumbersStep::makePipe()
...
         {
-            pipe.addSource(std::make_shared<NullSource>(NumbersSource::createHeader()));
+            pipe.addSource(std::make_shared<NullSource>(NumbersSource::createHeader(numbers_storage.column_name)));
             return pipe;
         }
         const auto & limit_length = limit_length_and_offset.first;
         const auto & limit_offset = limit_length_and_offset.second;
 
         /// If intersected ranges is limited or we can pushdown limit.
-        if (!intersected_ranges.rbegin()->right.isPositiveInfinity() || should_pushdown_limit)
```
Comment (alesapin) [inline src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp:450]: Why we don't need this condition anymore?

**Review by alesapin — COMMENTED** [review]
Only nitpicks, otherwise LGTM

**Review by alesapin — APPROVED** [review]
